### PR TITLE
Make progress on LKT semantics

### PR DESCRIPTION
**Pull request by raph-amiard**

### CODE REVIEW

File: testsuite/tests/contrib/lkt_semantic/test.out [hunk 1]
```diff
@@ -331,3 +331,24 @@ Id   <RefId "arg" test.lkt:48:53-48:56>
 Expr <NumLit test.lkt:48:58-48:60>
      has type <StructDecl "Int" __prelude:1:10-1:23>
 
+Id   <RefId "Int" test.lkt:51:9-51:12>
+     references <StructDecl "Int" __prelude:1:10-1:23>
+
+Resolving type of expr <NumLit test.lkt:51:15-51:17> raised an error
+    no type
```
Comment (Roldak) [inline testsuite/tests/contrib/lkt_semantic/test.out]: why does `12` have no type?

File: contrib/lkt/language/parser.py [hunk 2]
```diff
@@ -253,10 +253,16 @@ def in_type_ref():
 
     @langkit_property()
     def designated_scope():
-        # NOTE: for the moment, every dotted expression has a typed value on
-        # the RHS, and we'll use the type to go search valid fields and methods
-        # on it. We need to add support for packages/namespaces someday.
-        return Entity.expr_type.expr_type._.type_scope
+        ref_decl = Var(Entity.referenced_decl)
+
+        # If the RHS is a type, then use the type's scope to lookup operations
```
Comment (Roldak) [inline contrib/lkt/language/parser.py]: Why do you talk about `RHS` and not `Self`?

File: contrib/lkt/language/parser.py [hunk 3]
```diff
@@ -456,6 +456,12 @@ def expr_type_impl(expected_type=T.TypeDecl.entity,
           you can still override ``expr_type_impl``. This should not be
           necessary though, and should only be done following a discussion with
           LKT devs, and altering this comment afterwards.
+
+        NOTE: In the absence of a solver based more clever type inference
+        system (for now at least), we have to make a decision about the
+        direction types commonly flow in LKT. What follows from the previous
+    r   description is that type commonly flow upwards, from parent nodes to
```
Comment (Roldak) [inline contrib/lkt/language/parser.py]: weird `r`

File: contrib/lkt/language/parser.py (hunk 2 above)
Comment (raph-amiard) [reply contrib/lkt/language/parser.py]: You're correct, fixed

**Pull request merged**